### PR TITLE
Unify linux assembly files

### PR DESCRIPTION
**Pull request by lundman**
Make both Linux and Windows use the same assembler files. 

This is only the Linux part of the work, the Windows changes are here:
https://github.com/openzfsonwindows/openzfs/pull/188/commits

### Motivation and Context

Avoiding code-duplication is sometimes desirable, the question is, is
this one of those times?

### Description

Turns out clang (gnu) has a way to tell C to call an assembler function
using sysv_abi (and handle the register parameter order, and pushing
on-volatile registers on the stack) on Windows, thereby, making it
possible to use the same Unix style assembler files.

Linux will have to accept some changes, like always using ENTRY_NP and
similar macros.

I believe macOS could also be done, as I consider Windows platform to be the
most "foreign" to the Unix way.

It's not complete, this only handles x86_64. But in future, ARM64 will probably also
attempted. `ASMABI` is #define-d as empty initially, for all the non-x86 platforms, and
changed inside #ifdef x86. The other way to do it, is to list all platforms in an #ifdef  each.

I moved `asm_linkage.h` from `module/icp` to `include/spl` purely for `lua/setjmp` (the only .S file outside icp). By far the hardest part of this PR. But that could be beneficial for future .S work.

Please advise if this is desirable at all.

I don't care about the macro names, so specify what is better (name space?)

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by lundman** [comment]
Oh rats, need to figure out how Linux adds header locations.

File: module/icp/asm-x86_64/sha2/sha512_impl.S [hunk 1]
```diff
@@ -2110,6 +2110,6 @@ K512:
 	.quad	0x5fcb6fab3ad6faec,0x6c44198c4a475817
 #endif /* !lint && !__lint */
 
-#ifdef __ELF__
+#if defined(__linux__) && defined(__ELF__)
 .section .note.GNU-stack,"",%progbits
```
[the remark this answers is not in the document]
Comment (lundman) [reply module/icp/asm-x86_64/sha2/sha512_impl.S:2114]: Unsure now - half the assembler files have linux and the other doesn't.

**Comment by lundman** [comment]
Sheesh getting the includes to work after moving the header file is a pain. It's mostly because there is 1 set of assembler files (setjmp) in ZFS.

File: module/icp/algs/blake3/blake3_impl.c [hunk 2]
```diff
@@ -271,7 +271,7 @@ blake3_per_cpu_ctx_fini(void)
 
 #define	IMPL_FMT(impl, i)	(((impl) == (i)) ? "[%s] " : "%s ")
 
-#if defined(__linux__)
+#if defined(__linux__) || defined(_WIN32)
```
Comment (lundman) [inline module/icp/algs/blake3/blake3_impl.c:274]: Technically, this should be in the windows part of the PR

**Comment by lundman** [comment]
It's ready, the two failures is probably unrelated.

File: include/os/freebsd/spl/sys/ia32/asm_linkage.h [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+x:; \
+y:
+
+
+/*
+ * SET_SIZE trails a function and set the size for the ELF symbol table.
+ */
+#define	SET_SIZE(x)
+
+#define	SET_OBJ(x)
```
Comment (ryao) [inline include/os/freebsd/spl/sys/ia32/asm_linkage.h:163]: Why is this not defined for FreeBSD when the current code is equivalent to defining it for FreeBSD?

**Review by ryao — CHANGES_REQUESTED** [review]
Please change the commit message to say "assembly files". We do not call our C files "compiler files", so saying "assembler files" is inconsistent.

File: lib/libspl/include/os/freebsd/sys/ia32/asm_linkage.h [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+/*
+ * Copyright 2008 Sun Microsystems, Inc.  All rights reserved.
+ * Use is subject to license terms.
+ */
+
+#ifndef _IA32_SYS_ASM_LINKAGE_H
+#define	_IA32_SYS_ASM_LINKAGE_H
+
+#if defined(__linux__) && defined(CONFIG_SLS)
```
Comment (ryao) [inline lib/libspl/include/os/freebsd/sys/ia32/asm_linkage.h:30]: This should be in `lib/libspl/include/os/linux/sys/ia32/asm_linkage.h`. This is both the wrong platform. It is also in userspace where `CONFIG_SLS` is never defined.

That being said, would it be terrible to do this unconditionally everywhere in the interest of preemptive security? It is just dead code and honestly, our binary sizes are not going to be significantly influenced by this.

**Review by ryao — COMMENTED** [review]
This seems to have not given any attention to our ARM and POWER assembly. Was that intentional?

File: include/os/freebsd/spl/sys/ia32/asm_linkage.h (hunk 3 above)
Comment (lundman) [reply include/os/freebsd/spl/sys/ia32/asm_linkage.h:163]: Good catch!

File: lib/libspl/include/os/freebsd/sys/ia32/asm_linkage.h (hunk 4 above)
Comment (lundman) [reply lib/libspl/include/os/freebsd/sys/ia32/asm_linkage.h:30]: I copied the original (linux) over, so it certainly can be cleaned up for freebsd.  Hopefully the Linux versions are "untouched" beyond what I added.

**Comment by lundman** [comment]
> Please change the commit message to say "assembly files". We do not call our C files "compiler files", so saying "assembler files" is inconsistent.

Somewhat interesting linguistically, when C64 came to Europe, we are started calling it assembler for everything, to the point where `C64-wiki` as to start with _Assembly language, also incorrectly referred to as assembler_. But that's probably mostly down to ESOL. It's certainly stuck for me.

**Comment by ryao** [comment]
> > Please change the commit message to say "assembly files". We do not call our C files "compiler files", so saying "assembler files" is inconsistent.
> 
> Somewhat interesting linguistically, when C64 came to Europe, we are started calling it assembler for everything, to the point where `C64-wiki` as to start with _Assembly language, also incorrectly referred to as assembler_. But that's probably mostly down to ESOL. It's certainly stuck for me.

That explains why I heard that the translation into Polish was literally "assembler".

File: include/os/linux/spl/sys/ia32/asm_linkage.h [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+#endif /* <cet.h> */
+#endif /* __ELF__ && __CET__ && __has_include */
+#endif /* !ENDBR */
+
+#ifndef ENDBR
+#define	ENDBR
+#endif
+#ifndef RET
+#define	RET	ret
```
Comment (ryao) [inline include/os/linux/spl/sys/ia32/asm_linkage.h:53]: Do FreeBSD and Linux need to have different asm_linkage.h files?

I feel like we could have 1 file in a path such as `include/os/posix/spl/sys/ia32/asm_linkage.h`.

File: include/os/linux/spl/sys/ia32/asm_linkage.h [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+#define	A16	addr16;
+
+/*
+ * (There are some weird constructs in constant expressions)
+ */
+#define	_CONST(const)		[const]
+#define	_BITNOT(const)		-1!_CONST(const)
+#define	_MUL(a, b)		_CONST(a \* b)
+
+#else
```
Comment (ryao) [inline include/os/linux/spl/sys/ia32/asm_linkage.h:91]: A comment `/* __GNUC_AS__ */` following #else on the same line would make this more readable.

File: include/os/linux/spl/sys/ia32/asm_linkage.h [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+ * us about what we're trying to do.
+ */
+#define	D16	.byte	0x66;
+#define	A16	.byte	0x67;
+
+#define	_CONST(const)		(const)
+#define	_BITNOT(const)		~_CONST(const)
+#define	_MUL(a, b)		_CONST(a * b)
+
+#endif
```
Comment (ryao) [inline include/os/linux/spl/sys/ia32/asm_linkage.h:104]: A comment `/* __GNUC_AS__ */` following #endif on the same line would make this more readable.

File: include/os/linux/spl/sys/ia32/asm_linkage.h (hunk 5 above)
Comment (lundman) [reply include/os/linux/spl/sys/ia32/asm_linkage.h:53]: You probably could, but I believe I have to be different for macOS already. We have OS separation, have a posix/macos/windows separation would be something new.

**Comment by andrewc12** [comment]
@lundman hows this coming along?

**Review by behlendorf — APPROVED** [review]
@lundman if you can rebase this I believe it should be ready to merge.

**Comment by lundman** [comment]
rebased.

**Comment by lundman** [comment]
Does the test
`Test: /usr/share/zfs/zfs-tests/tests/functional/pyzfs/pyzfs_unittest (run as root) [18:46] [KILLED]`

have anything to do with crypto, is it something to look into?

**Comment by behlendorf** [comment]
It looks like a previous test failure prevented `pyzfs_unittest` from running properly.  It doesn't look like it had anything to do with crypto, I'll resubmit the Ubuntu 20.04 build.

**Review by ryao — APPROVED** [review]

File: module/icp/asm-x86_64/modes/ghash-x86_64.S [hunk 8]
```diff
@@ -155,12 +156,10 @@ gcm_gmult_clmul:
...
 	RET
 .cfi_endproc
-.size	gcm_gmult_clmul,.-gcm_gmult_clmul
+SET_SIZE(gcm_gmult_clmul)
+#endif /* !_WIN32 || _KERNEL */
 
-.globl	gcm_init_htab_avx
-.type	gcm_init_htab_avx,@function
-.align	32
-gcm_init_htab_avx:
```
Comment (AttilaFueloep) [inline module/icp/asm-x86_64/modes/ghash-x86_64.S:163]: alignment change 32 -> 16

File: module/icp/asm-x86_64/modes/ghash-x86_64.S [hunk 9]
```diff
@@ -269,21 +268,17 @@ gcm_init_htab_avx:
 	vzeroupper
 	RET
 .cfi_endproc
-.size	gcm_init_htab_avx,.-gcm_init_htab_avx
+SET_SIZE(gcm_init_htab_avx)
 
-.globl	gcm_gmult_avx
-.type	gcm_gmult_avx,@function
-.align	32
-gcm_gmult_avx:
```
Comment (AttilaFueloep) [inline module/icp/asm-x86_64/modes/ghash-x86_64.S:277]: alignment change 32 -> 16

File: module/icp/asm-x86_64/modes/ghash-x86_64.S [hunk 10]
```diff
@@ -269,21 +268,17 @@ gcm_init_htab_avx:
...
+ENTRY_NP(gcm_gmult_avx)
 .cfi_startproc
 	ENDBR
 	jmp	.L_gmult_clmul
 .cfi_endproc
-.size	gcm_gmult_avx,.-gcm_gmult_avx
-.globl	gcm_ghash_avx
-.type	gcm_ghash_avx,@function
-.align	32
-gcm_ghash_avx:
```
Comment (AttilaFueloep) [inline module/icp/asm-x86_64/modes/ghash-x86_64.S:286]: alignment change 32 -> 16

File: module/lua/setjmp/setjmp_x86_64.S [hunk 11]
```diff
@@ -62,7 +50,7 @@ x:
 	RET
 	SET_SIZE(setjmp)
 
-	ENTRY(longjmp)
```
Comment (AttilaFueloep) [inline module/lua/setjmp/setjmp_x86_64.S:65]: alignment change 8 -> 16

File: module/lua/setjmp/setjmp_x86_64.S [hunk 12]
```diff
@@ -27,28 +27,16 @@
...
-#define	SET_SIZE(x) \
-	.size	x, [.-x]
-
 /*
  * Setjmp and longjmp implement non-local gotos using state vectors
  * type label_t.
  */
 #ifdef __x86_64__
 
-	ENTRY(setjmp)
```
Comment (AttilaFueloep) [inline module/lua/setjmp/setjmp_x86_64.S:51]: alignment change 8 -> 16

**Review by AttilaFueloep — CHANGES_REQUESTED** [review]
Sorry for being late to the party again but unfortunately I'd prefer to see some changes made.

By introducing `ENTRY_NP` in some asm files the code alignment of some functions is changed from 32 or 64 bytes to 16 bytes. While this won't break the code, it may degrade performance noticeably or even considerably.

Off the top of my head, Haswells instructions fetcher operates on 32 byte blocks. Not sure which architecture would profit from 64 byte alignment, but 16 byte alignment would definitely not be large enough in some cases.

Given the highly optimized assembly and the complexity of code alignment, I'd prefer to stay on the safe side and keep the original alignments. This could be done by introducing a new macro, say `ENTRY_NP_WITH_ALIGN`, which takes an additional alignment argument.

Regarding the alignment change (8 -> 16) in `lua/setjmp`, I've no idea if this will have any impact on functionality but if you add the above macro it won't hurt to keep the original alignment here as well.

Any thoughts?

**Comment by lundman** [comment]
Yeah, this is a good point. We could take align out of ENTRY_NP() and keep it separate, or as you suggest, add it to macro. But since it's an illumos macro, perhaps separate is better.
One annoying thing is, sometimes it's literal "16" and sometimes its bit shift "4" - but perhaps the macro can handle that.

File: module/icp/algs/blake3/blake3_x86-64.c [hunk 13]
```diff
@@ -29,20 +29,20 @@
...
-extern void zfs_blake3_compress_xof_sse2(const uint32_t cv[8],
+extern void ASMABI zfs_blake3_compress_xof_sse2(const uint32_t cv[8],
     const uint8_t block[BLAKE3_BLOCK_LEN], uint8_t block_len,
     uint64_t counter, uint8_t flags, uint8_t out[64]);
 
-extern void zfs_blake3_hash_many_sse2(const uint8_t * const *inputs,
+extern void ASMABI zfs_blake3_hash_many_sse2(const uint8_t * const *inputs,
     size_t num_inputs, size_t blocks, const uint32_t key[8],
     uint64_t counter, boolean_t increment_counter, uint8_t flags,
     uint8_t flags_start, uint8_t flags_end, uint8_t *out);
```
Comment (lundman) [inline module/icp/algs/blake3/blake3_x86-64.c:44]: below is bad, doesn't affect linux, but on windows, you better believe there's fireworks! only assembler functions should be marked sysv :)

**Comment by AttilaFueloep** [comment]
> We could take align out of ENTRY_NP() and keep it separate

`ENTRY_NP()` is used in other places as well IIRC, my suggestion was to do something like

```c
#define	ENTRY_NP_WITH_ALIGN(x,a) \
	.text; \
	.align	a; \
	.globl	x; \
x:
```
and use it in places where the alignment isn't 16.

> One annoying thing is, sometimes it's literal "16" and sometimes its bit shift "4"

GAS has two directives, `.align <bytes>` and `.p2align  <power of two>`. IIRC llvm only has `.p2align`, so maybe it would be better to use `.p2align`. Both take two [additional arguments](https://ftp.gnu.org/old-gnu/Manuals/gas-2.9.1/html_chapter/as_toc.html#TOC70), the first of which is the fill value. In the text segment this would be `NOP` per default, and  the opcode of `NOP`  is 0x90. So `.align 16` and `.align 16, 0x90` are equivalent.

> `.align 6, 0x90`

This syntax is new to me, did you mean `.p2align 6, 0x90` by any chance? Or is this something Windows specific?

File: module/icp/algs/blake3/blake3_x86-64.c (hunk 13 above)
Comment (AttilaFueloep) [reply module/icp/algs/blake3/blake3_x86-64.c:44]: Sure, if you call a Windows compiled C function with sysv semantics all arguments will be in the wrong places.

**Comment by AttilaFueloep** [comment]
Well, actually `.align 6` is llvm syntax and would translate to `.align 64` in GAS. The good new is that llvm also has a `.p2align`  directive, so the Mac port would profit from using `.p2align` everywhere.

**Comment by lundman** [comment]
macOS has the 2 argument align, you specify the value you want in there, 0x90 is NOP or something.
https://github.com/openzfsonosx/openzfs/blob/development/module/icp/asm-x86_64/os/macos/aes/aes_amd64.S#L696

but I see I've left the align problem alone for now, when this PR lands, I'll do a PR for macOS, to handle the align macro.

If one of align, or p2align, works on macOS, I'll just change it to that, otherwise a new MACRO

**Comment by lundman** [comment]
OK so, I don't see anything using `MCOUNT`, whatever that is. I think maybe it was something `mdb` related for illumos kernels? Either way, we are not using it.  (ok, GPROF)
https://github.com/illumos/illumos-gate/blob/9ecd05bdc59e4a1091c51ce68cce2028d5ba6fd1/usr/src/uts/sparc/sys/asm_linkage.h#L66-L76
So if we are moving away from illumos macros, I just went with `ENTRY_ALIGN()` macro. Lemme know if anyone have strong opinions on names.

File: module/icp/asm-x86_64/modes/aesni-gcm-x86_64.S [hunk 14]
```diff
@@ -50,14 +50,16 @@
...
 .extern gcm_avx_can_use_movbe
 
 .text
 
 #ifdef HAVE_MOVBE
-.type	_aesni_ctr32_ghash_6x,@function
-.align	32
-_aesni_ctr32_ghash_6x:
+.align 32
+FUNCTION(_aesni_ctr32_ghash_6x)
```
Comment (AttilaFueloep) [inline module/icp/asm-x86_64/modes/aesni-gcm-x86_64.S:62]: ENTRY_ALIGN()? Looks like an old version?

File: module/icp/asm-x86_64/modes/aesni-gcm-x86_64.S [hunk 15]
```diff
@@ -369,12 +371,11 @@ _aesni_ctr32_ghash_6x:
...
 .cfi_endproc
-.size	_aesni_ctr32_ghash_6x,.-_aesni_ctr32_ghash_6x
+SET_SIZE(_aesni_ctr32_ghash_6x)
 #endif /* ifdef HAVE_MOVBE */
 
-.type	_aesni_ctr32_ghash_no_movbe_6x,@function
-.align	32
-_aesni_ctr32_ghash_no_movbe_6x:
+.align 32
+FUNCTION(_aesni_ctr32_ghash_no_movbe_6x)
```
Comment (AttilaFueloep) [inline module/icp/asm-x86_64/modes/aesni-gcm-x86_64.S:378]: As above.

File: module/icp/asm-x86_64/modes/aesni-gcm-x86_64.S [hunk 16]
```diff
@@ -818,10 +816,10 @@ aesni_gcm_decrypt:
...
 	RET
 .cfi_endproc
-.size	aesni_gcm_decrypt,.-aesni_gcm_decrypt
-.type	_aesni_ctr32_6x,@function
-.align	32
-_aesni_ctr32_6x:
+SET_SIZE(aesni_gcm_decrypt)
+
+.align 32
+FUNCTION(_aesni_ctr32_6x)
```
Comment (AttilaFueloep) [inline module/icp/asm-x86_64/modes/aesni-gcm-x86_64.S:822]: As above.

**Review by AttilaFueloep — CHANGES_REQUESTED** [review]
Looks like an old version sneaked in partially.

File: module/icp/asm-x86_64/modes/aesni-gcm-x86_64.S (hunk 14 above)
Comment (lundman) [reply module/icp/asm-x86_64/modes/aesni-gcm-x86_64.S:62]: This is a special case, the function `aesni_ctr32_ghash_6x` in not called from C, so this PR doesn't need to `ASMABI` it, and it is not defined as `globl` so no need for `ENTRY_NP`.

it is defined as:

```
.type	_aesni_ctr32_ghash_6x,@function
.align	32
_aesni_ctr32_ghash_6x:
```

and Windows don't handle the `.type function` so I made a macro `FUNCTION()` to put that in for Linux, and empty for Windows.
I left the `align` as if for now, when I do macOS, I'll do a PR to unify the aligns.

Comment (AttilaFueloep) [reply module/icp/asm-x86_64/modes/aesni-gcm-x86_64.S:62]: Indeed, I should've noticed it. Sorry for the noise.

**Review by AttilaFueloep — APPROVED** [review]

**Comment by AttilaFueloep** [comment]
@lundman  When you do macOS  you could try `.balign` first. It has the least impact (same syntax as `.align` on Linux) and should work fine wit one argument on macOS.  I tried it with GAS and clang on Linux, and they behave identical,  padding with nops inside the .text segment and with zeros inside the .data segment .

**Comment by lundman** [comment]
>  try .balign first.

Will do.

@behlendorf  I believe we are ready.

**Comment by behlendorf** [comment]
@lundman @AttilaFueloep thanks for the additional review and refactoring!  I'll get this pulled in.

**Pull request merged**